src/OrchardCore.Modules/OrchardCore.AdminTrees/Controllers/NodeController.cs

```diff
@@ -0,0 +1,349 @@
+            var model = new AdminNodeEditViewModel
+            {
+                AdminTreeId = id,
+                AdminNode = treeNode,
+                AdminNodeId = treeNode.UniqueId,
+                AdminNodeType = treeNode.GetType().Name,
+                Editor = await _displayManager.BuildEditorAsync(treeNode, updater: this, isNew: false)
+            };
+
+            model.Editor.TreeNode = treeNode;
+
+            return View(model);
+        }
+
+        [HttpPost]
+        public async Task<IActionResult> Edit(AdminNodeEditViewModel model)
+        {
+            if (!await _authorizationService.AuthorizeAsync(User, Permissions.ManageAdminTree))
+            {
+                return Unauthorized();
+            }
+
+            var tree = await _session.GetAsync<AdminTree>(model.AdminTreeId);
+
+            if (tree == null)
+            {
+                return NotFound();
+            }
+
+            var treeNode = tree.GetMenuItemById(model.AdminNodeId);
+
+            if (treeNode == null)
+            {
+                return NotFound();
+            }
+
+            var editor = await _displayManager.UpdateEditorAsync(treeNode, updater: this, isNew: false);
+
+            if (ModelState.IsValid)
+            {
+                _session.Save(tree);
+
+                _notifier.Success(H["Admin node updated successfully"]);
+                return RedirectToAction(nameof(List), new { id = model.AdminTreeId});
+            }
+
+            _notifier.Error(H["The admin node has validation errors"]);
+            model.Editor = editor;
+
+            // If we got this far, something failed, redisplay form
+            return View(model);
+        }
+
+        [HttpPost]
+        public async Task<IActionResult> Delete(int id, string treeNodeId)
+        {
+            if (!await _authorizationService.AuthorizeAsync(User, Permissions.ManageAdminTree))
+            {
+                return Unauthorized();
+            }
+
+            var tree = await _session.GetAsync<AdminTree>(id);
+
+            if (tree == null)
+            {
+                return NotFound();
+            }
+
+            var treeNode = tree.GetMenuItemById(treeNodeId);
+
+            if (treeNode == null)
+            {
+                return NotFound();
+            }
+
+            if (tree.RemoveMenuItem(treeNode) == false)
+            {
+                return new StatusCodeResult(500);
+            }
+            
+            
+            _session.Save(tree);
+
+            _notifier.Success(H["Admin node deleted successfully"]);
+
+            return RedirectToAction(nameof(List), new { id });
+        }
+
+        [HttpPost]
+        public async Task<IActionResult> Toggle(int id, string treeNodeId)
+        {
+            if (!await _authorizationService.AuthorizeAsync(User, Permissions.ManageAdminTree))
+            {
+                return Unauthorized();
+            }
+
+            var tree = await _session.GetAsync<AdminTree>(id);
```

Comment:
This should go in a service class, and be cached and invalidated when it changes. Look at the site settings for an example on how to cache it in memory.

Reply:
Done